### PR TITLE
feat(ELB): add elb security policies data source

### PR DESCRIPTION
**Pull request by SabreKeyZ**
**What this PR does / why we need it**:
add elb security policies data source
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST=./huaweicloud/services/acceptance/elb/ TESTARGS='-run TestAccDatasourceSecurityPolicies'        
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/elb/ -v -run TestAccDatasourceSecurityPolicies -timeout 360m -parallel 4 
=== RUN   TestAccDatasourceSecurityPolicies_basic 
=== PAUSE TestAccDatasourceSecurityPolicies_basic
=== CONT  TestAccDatasourceSecurityPolicies_basic
--- PASS: TestAccDatasourceSecurityPolicies_basic (12.78s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/elb       12.816s

```

### CODE REVIEW

File: docs/data-sources/elb_security_policies.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String) Specifies the region in which to query the data source.
+  If omitted, the provider-level region will be used.
+
+* `name` - (Optional, String) Specifies the name of the custom/system security policy.
+
+* `type` - (Optional, String) Specifies the type of the security policy.
```
Comment (houpeng80) [inline docs/data-sources/elb_security_policies.md]:
```suggestion
* `type` - (Optional, String) Specifies the type of the security policy. Value options: **system**, **custom**.
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+		}
+		if descriptionExist && securityPoliciesdescription.(string) != description.(string) {
+			continue
+		}
+		if protocolsExist && !filterQueryParams(protocols, securityPoliciesProtocols.(string)) {
+			continue
+		}
+		if ciphersExist && !filterQueryParams(ciphers, securityPoliciesCiphers.(string)) {
+			continue
+		}
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
		if protocolsExist && !utils.StrSliceContains(protocols, securityPoliciesProtocol.(string)) {
			continue
		}
		if ciphersExist && !utils.StrSliceContains(ciphers, securityPoliciesCipher.(string)) {
			continue
		}
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+	}
+	curJson := utils.PathSearch("system_security_policies", resp, make([]interface{}, 0))
+	if curJson == nil {
+		return nil
+	}
+	securityPoliciesName, nameExist := d.GetOk("name")
+	securityPoliciesProtocols, protocolsExist := d.GetOk("protocol")
+	securityPoliciesCiphers, ciphersExist := d.GetOk("cipher")
+	securityPoliciesId, idExist := d.GetOk("security_policies_id")
+	securityPoliciesdescription, descriptionExist := d.GetOk("description")
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
	securityPoliciesName, nameOk := d.GetOk("name")
	securityPoliciesProtocols, protocolsOk := d.GetOk("protocol")
	securityPoliciesCiphers, ciphersOk := d.GetOk("cipher")
	securityPoliciesId, idOk := d.GetOk("security_policies_id")
	securityPoliciesdescription, descriptionOk := d.GetOk("description")
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+	}
+	return listSystemSecurityPoliciesRespBody, nil
+}
+
+func customElbSecurityPolicies(d *schema.ResourceData, meta interface{}) (interface{}, error) {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	var (
+		listSecurityPoliciesHttpUrl = "v3/{project_id}/elb/security-policies"
+		listSecurityPoliciesProduct = "elb"
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
		listCustomSecurityPoliciesHttpUrl = "v3/{project_id}/elb/security-policies"
		listCustomSecurityPoliciesProduct = "elb"
```
and fix other places

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+	listSystemSecurityPoliciesPath = strings.ReplaceAll(listSystemSecurityPoliciesPath, "{project_id}",
+		listSystemSecurityPoliciesClient.ProjectID)
+	getSystemSecurityPoliciesOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+
+	listSystemSecurityPoliciesResp, err := listSystemSecurityPoliciesClient.Request("GET",
+		listSystemSecurityPoliciesPath, &getSystemSecurityPoliciesOpt)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving ELB security policies")
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
		return nil, fmt.Errorf("error retrieving ELB system security policies")
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+		listSecurityPoliciesClient.ProjectID)
+	listSecurityPoliciesQueryParams := buildListSecurityPoliciesQueryParams(d)
+	listSecurityPoliciesPath += listSecurityPoliciesQueryParams
+	listSecurityPoliciesResp, err := pagination.ListAllItems(
+		listSecurityPoliciesClient,
+		"marker",
+		listSecurityPoliciesPath,
+		&pagination.QueryOpts{MarkerField: ""})
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving ELB security policies")
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
		return nil, fmt.Errorf("error retrieving ELB custom security policies")
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+	if v, ok := d.GetOk("description"); ok {
+		res = fmt.Sprintf("%s&description=%v", res, v)
+	}
+	if res != "" {
+		res = "?" + res[1:]
+	}
+	return res
+}
+
+func flattenSecurityPoliciesBody(resp interface{}, d *schema.ResourceData) []interface{} {
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
func flattenCustomSecurityPoliciesBody(resp interface{}, d *schema.ResourceData) []interface{} {
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+	if resp == nil {
+		return nil
+	}
+	curJson := utils.PathSearch("security_policies", resp, make([]interface{}, 0))
+	if curJson == nil {
+		return nil
+	}
+	securityPoliciesProtocol, protocolsExist := d.GetOk("protocol")
+	securityPoliciesCipher, ciphersExist := d.GetOk("cipher")
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
	securityPoliciesProtocol, protocolsOk := d.GetOk("protocol")
	securityPoliciesCipher, ciphersOk := d.GetOk("cipher")
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+	curArray := curJson.([]interface{})
+	rst := make([]interface{}, 0, len(curArray))
+	for _, v := range curArray {
+		protocols := utils.PathSearch("protocols", v, "")
+		ciphers := utils.PathSearch("ciphers", v, "")
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
		protocols := utils.PathSearch("protocols", v, make([]interface{}))
		ciphers := utils.PathSearch("ciphers", v, make([]interface{}))
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+		if protocolsExist && !filterQueryParams(getStrArray(protocols), securityPoliciesProtocol.(string)) {
+			continue
+		}
+		if ciphersExist && !filterQueryParams(getStrArray(ciphers), securityPoliciesCipher.(string)) {
+			continue
+		}
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
		if protocolsExist && !utils.StrSliceContains(utils.ExpandToStringList(protocols), securityPoliciesProtocol.(string))  {
			continue
		}
		if ciphersExist && !utils.StrSliceContains(utils.ExpandToStringList(ciphers), securityPoliciesCipher.(string)) {
			continue
		}
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+	securityPoliciesProtocols, protocolsExist := d.GetOk("protocol")
+	securityPoliciesCiphers, ciphersExist := d.GetOk("cipher")
+	securityPoliciesId, idExist := d.GetOk("security_policies_id")
+	securityPoliciesdescription, descriptionExist := d.GetOk("description")
+	curArray := curJson.([]interface{})
+	rst := make([]interface{}, 0, len(curArray))
+	for _, v := range curArray {
+		name := utils.PathSearch("name", v, "")
+		protocols := strings.Split(utils.PathSearch("protocols", v, "").(string), " ")
+		ciphers := strings.Split(utils.PathSearch("ciphers", v, "").(string), ":")
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
		protocols := utils.ExpandToStringList(strings.Split(utils.PathSearch("protocols", v, "").(string), "\\s"))
		ciphers := utils.ExpandToStringList(strings.Split(utils.PathSearch("ciphers", v, "").(string), ":"))
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+		}
+		if ciphersExist && !filterQueryParams(ciphers, securityPoliciesCiphers.(string)) {
+			continue
+		}
+		rst = append(rst, map[string]interface{}{
+			"type":      "system",
+			"id":        utils.PathSearch("id", v, nil),
+			"name":      utils.PathSearch("name", v, nil),
+			"protocols": getProtocols(v),
+			"ciphers":   getCiphers(v),
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
			"protocols": protocols,
			"ciphers":   ciphers,
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+func dataSourceElbSecurityPoliciesRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	v := d.Get("type").(string)
+	if v == "system" {
+		return systemElbSecurityPoliciesRead(d, meta)
+	} else if v == "custom" {
+		return customElbSecurityPoliciesRead(d, meta)
+	}
+	return allElbSecurityPoliciesRead(d, meta)
```
Comment (houpeng80) [inline huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]:
```suggestion
        cfg := meta.(*config.Config)
	region := cfg.GetRegion(d)
	var (
		listSecurityPoliciesProduct = "elb"
	)
	listSecurityPoliciesClient, err := cfg.NewServiceClient(lisSecurityPoliciesProduct, region)
	if err != nil {
		return nil, fmt.Errorf("error creating ELB client: %s", err)
	}
	var sceurityPolicies []interface{}
	v := d.Get("type").(string)
	if v == "system" {
		sceurityPolicies = systemElbSecurityPoliciesRead(d, listSecurityPoliciesClient, meta)
	} else if v == "custom" {
		sceurityPolicies = customElbSecurityPoliciesRead(d, listSecurityPoliciesClient, meta)
	}else{
	        sceurityPolicies = allElbSecurityPoliciesRead(d,listSecurityPoliciesClient, meta)
	}
	dataSourceId, err := uuid.GenerateUUID()
	if err != nil {
		return diag.Errorf("unable to generate ID: %s", err)
	}
	d.SetId(dataSourceId)

	mErr := multierror.Append(
		d.Set("region", region),
		d.Set("security_policies", sceurityPolicies),
	)
```

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 2 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 3 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 4 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 5 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 6 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 7 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 8 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 9 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 10 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 11 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 12 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go (hunk 13 above)
Comment (SabreKeyZ) [reply huaweicloud/services/elb/data_source_huaweicloud_elb_security_policies.go]: done

File: docs/data-sources/elb_security_policies.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+The following arguments are supported:
+
+* `region` - (Optional, String) Specifies the region in which to query the data source.
+  If omitted, the provider-level region will be used.
+
+* `name` - (Optional, String) Specifies the name of the security policy.
+
+* `type` - (Optional, String) Specifies the type of the security policy. Value options: **system**, **custom**.
+
+* `security_policies_id` - (Optional, String) Specifies the ID of the security policy.
```
Comment (ShiChangkuo) [inline docs/data-sources/elb_security_policies.md]: it should be **security_policy_id**

Comment (SabreKeyZ) [reply docs/data-sources/elb_security_policies.md]: done

**Pull request merged**